### PR TITLE
Assignment 6 - Paulson Ray Thompson & Trevor Stewart

### PR DESCRIPTION
**Pull request by Purplepigs0129**
Paulson's Git - @Purplepigs0129
Trevor's Git - @Tvor369

### CODE REVIEW

File: Calculate/Calculator.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    public static int Multiple(int a, int b)
+    {
+        int answer = a * b;
+        return answer;
+    }
+
+    public static int Divide(int a, int b)
+    {
+        if (b == 0)
```
Comment (SaaWest) [inline Calculate/Calculator.cs:36]: nice condition

File: Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                        result = null;
+                        return false;
+                    }
+
+                    Func<int, int, int> opperation = MathematicalOperations[opperand[0]];
+                    //result = opperation(double.Parse(equationParts[0]), double.Parse(equationParts[2]));
+                    result = opperation(variable1, variable2);
+                    return true;
+                }
+                catch (FormatException)
```
Comment (SaaWest) [inline Calculate/Calculator.cs:72]: nice addition for invalid entries

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            result = null;
+            return false;
+        }
+        else
+        {
+            if (int.TryParse(equationParts[0], out int variable1) && int.TryParse(equationParts[2], out int variable2))
+            {
+                try
+                {
+                    char[] opperand = equationParts[1].ToCharArray();
```
Comment (SaaWest) [inline Calculate/Calculator.cs:60]: why not write this as `char opperand = equationParts[1][0]`?

File: Calculate/Program.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace Calculate;
+public class Program
+{
+#nullable disable
+    public Action<string> WriteLine { get; init; } = Console.WriteLine;
+    public Func<string> ReadLine { get; init; } = Console.ReadLine;
```
Comment (SaaWest) [inline Calculate/Program.cs:8]: Nothing wrong really but why not `    public Func<string?> ReadLine { get; init; } = Console.ReadLine;` that way you dont have to trigger the #nullable

File: Calculate.Tests/ProgramTests.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    public void Program_DefaultConstructor_EqualsConsole()
+    {
+        Program program = new();
+
+        Assert.Equal(Console.WriteLine, program.WriteLine);
+        Assert.Equal(Console.ReadLine, program.ReadLine);
+    }
+
+    [Fact]
+    public void Main_UserGivesGoodInput_MatchExpected()
```
Comment (SaaWest) [inline Calculate.Tests/ProgramTests.cs:62]: nice use of the in class lecture test

File: Directory.Build.Props [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<Project>
```
Comment (SaaWest) [inline Directory.Build.Props:1]: needs to be Directory.Build.props not .Props. Ben talk about this after the class was done

**Review by SaaWest — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
    Define a Program Class
        Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
        Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
        Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
    Define a Calculator class ✔
        Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
        Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
            is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
                Uses char for the key corresponding to the operators +, -, *, and /. ✔
                Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
        Implement a TryCalculate method following "TryParse" pattern ✔
            Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
            If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
            Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
            Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
    Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
    Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

    Refactor the redirect portion of the Program class into 'ProgramBase`❌
    Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects❌
    Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)❌

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.✔ props needs to be .props
    Place all shared project items into a Directory.Build.targets file.❌
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔ but is in the csproj not Props file
    and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌ but i think the .Equals() is permissable
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                }
+                catch (FormatException)
+                {
+                    throw new FormatException($"Unable to Parse {equationParts[0]} or {equationParts[2]}");
+                }
+
+            }
+        }
+        result = null;
+        return false;
```
Comment (NathanMarsee) [inline Calculate/Calculator.cs:80]: This whole method is very well thought out. Using a large try statement takes a lot fewer lines than my solution took, good job!

File: Calculate/Program.cs (hunk 4 above)
Comment (NathanMarsee) [reply Calculate/Program.cs:8]: I second this, disabling nullability is not necessary.

File: Directory.Build.Props (hunk 6 above)
Comment (NathanMarsee) [reply Directory.Build.Props:1]: Nice catch, I wouldn't have noticed this in my review if you hadn't pointed it out.

File: Calculate/Calculator.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        int answer = a - b;
+        return answer;
+    }
+
+    public static int Multiple(int a, int b)
+    {
+        int answer = a * b;
+        return answer;
+    }
```
Comment (NathanMarsee) [inline Calculate/Calculator.cs:32]: Very minor thing, but you might want to consider condensing each of these three methods (Add, Subtract, Multiple) into single lines using the format `public static double Add(int a, int b) => a + b;`

File: Calculate.Tests/CalculatorTests.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+public class CalculatorTests
+{
+    [Theory]
+    [InlineData(4, 2, 2)]
+    [InlineData(9, 6, 3)]
+    [InlineData(-3, -5, 2)]
+    public void Add_ReturnSumOfTwoValues_Equal(int expected, int val1, int val2)
+    {
+        int sum = Calculator.Add(val1, val2);
+        Assert.Equal(expected, sum);
```
Comment (NathanMarsee) [inline Calculate.Tests/CalculatorTests.cs:15]: We are meant to use `Assert.AreEqual<T>()` (the generic version) for our equality tests.

**Review by NathanMarsee — COMMENTED** [review]
**Instructions:**
- Create a Console project called "Calculate". ✔
- Define a Program Class
   - Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
   - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
   - Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
   - Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
   - Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
      - is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
         - Uses char for the key corresponding to the operators +, -, *, and /. ✔
         - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
   - Implement a TryCalculate method following "TryParse" pattern ✔
      - Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
      - If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
      - Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
      - Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

**Extra Credit:**
Not Attempted

**Fundamentals**
- Place all shared project properties into a Directory.Build.Props file. ❌ _(small typo)_
- Place all shared project items into a Directory.Build.targets file. _(unnecessary)_
- nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔ 
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔


**Reviewer's Notes:**
Well done overall! there are just a couple minor things that could be fixed for the fundamentals category, and a couple ways you could shorten/simplify your code, but those are very minor issues. Nice work.

File: Calculate.Tests/ProgramTests.cs (hunk 5 above)
Comment (Purplepigs0129) [reply Calculate.Tests/ProgramTests.cs:62]: Thanks!

File: Calculate/Calculator.cs (hunk 1 above)
Comment (Purplepigs0129) [reply Calculate/Calculator.cs:36]: Thanks!

File: Calculate/Calculator.cs (hunk 2 above)
Comment (Purplepigs0129) [reply Calculate/Calculator.cs:72]: Thanks!

File: Calculate/Calculator.cs (hunk 7 above)
Comment (Purplepigs0129) [reply Calculate/Calculator.cs:80]: Thanks!

File: Directory.Build.Props (hunk 6 above)
Comment (Purplepigs0129) [reply Directory.Build.Props:1]: Your totally correct will fix

File: Calculate/Calculator.cs (hunk 8 above)
Comment (Purplepigs0129) [reply Calculate/Calculator.cs:32]: Will do. Got confused on the assignment when it says each of the methods take in 2 params and return a third. Much more simple thanks

File: Calculate/Calculator.cs (hunk 3 above)
Comment (Purplepigs0129) [reply Calculate/Calculator.cs:60]: Totally overshot this in my head thanks for the simplification

File: Calculate.Tests/CalculatorTests.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    public void Divide_ReturnQuotientOfTwoValues_Equal(int expected, int val1, int val2)
+    {
+        int quotient = Calculator.Divide(val1, val2);
+        Assert.Equal(expected, quotient);
+    }
+
+    [Fact]
+    public void Divide_DivideByZero_ThrowException()
+    {
+        Assert.Throws<ArgumentException>(() => Calculator.Divide(69, 0));
```
Comment (BenjaminMichaelis) [inline Calculate.Tests/CalculatorTests.cs:49]: Divide by zero by default throws a `System.DivideByZeroException`

File: Calculate/Calculator.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    };
+    public static int Add(int a, int b) => a + b; 
+
+    public static int Subtract(int a, int b) => a - b;
+
+    public static int Multiple(int a, int b) => a * b;
+
+    public static int Divide(int a, int b)
+    {
+        if (b == 0)
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:24]: Divide by zero by default throws a `System.DivideByZeroException`

File: Calculate/Program.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace Calculate;
+public class Program
+{
+#nullable disable
```
Comment (BenjaminMichaelis) [inline Calculate/Program.cs:6]: We should never disable nullable

File: Calculate.Tests/CalculatorTests.cs (hunk 9 above)
Comment (NathanMarsee) [reply Calculate.Tests/CalculatorTests.cs:15]: My bad, I haven't really looked into much Xunit at all and haven't checked Teams often enough. You're totally right on this.

**Pull request closed without merging**